docs/source/more_types.rst

```diff
@@ -933,3 +933,210 @@ and non-required keys, such as ``Movie`` above, will only be compatible with
+
+* There can be *at most one* final declaration per module or class for
+  a given attribute:
+
+  .. code-block:: python
+
+     ID: Final = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         id: Final = 1
+         def __init__(self, x: int) -> None:
+             self.id: Final = x  # Error!
+
+  Note that mypy has a single namespace for a class. So there can't be two
+  class-level and instance-level constants with the same name.
+
+* There must be *exactly one* assignment to a final attribute:
+
+  .. code-block:: python
+
+     ID = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         ID = 1
+         ID: Final = 2  # Error!
+
+.. note::
+   Conditional final declarations and final declarations within loops are
+   not supported.
+
+Using final attributes
+----------------------
+
+As a result of a final declaration mypy strives to provide the
+two following guarantees:
+
+* A final attribute can't be re-assigned (or otherwise re-defined), both
+  internally and externally:
+
+  .. code-block:: python
+
+     # file mod.py
+     from typing import Final
+
+     ID: Final = 1
+
+     # file main.py
+     from typing import Final
+
+     import mod
+     mod.ID = 2  # Error, can't assign to constant.
+
+     class SomeCls:
+         ID: Final = 1
+
+         def meth(self) -> None:
+             self.ID = 2  # Error, can't assign to final attribute
+
+     class DerivedCls(SomeCls):
+         ...
+
+     DerivedCls.ID = 2  # Error!
+     obj: DerivedCls
+     obj.ID = 2  # Error!
+
+* A final attribute can't be overridden by a subclass (even with another
+  explicit final declaration). Note however, that final attributes can
+  override normal attributes. This also applies to multiple inheritance:
+
+  .. code-block:: python
+
+     class Base:
+         ID = 0
+
+     class One(Base):
+         ID: Final = 1  # OK
+
+     class Other(Base):
+         ID: Final = 2  # OK
+
+     class Combo(One, Other):  # Error, cannot override final attribute.
+         pass
+
+Final methods
+-------------
+
+Like with attributes, sometimes it is useful to protect a method from
+overriding. In such situations one can use a ``typing.final`` decorator:
+
+.. code-block:: python
+
+   from typing import final
+
+   class Base:
+       @final
+       def common_name(self) -> None:  # common signature
+           ...
+
+   # 1000 lines later
+
+   class Derived(Base):
+       def common_name(self) -> None:  # Error, this overriding might break
+                                       # invariants in the base class.
+           ...
+
+This ``@final`` decorator can be used with instance methods, class methods,
+static methods, and properties (this includes overloaded methods).
+
+Final classes
+-------------
+
+As a bonus, applying a ``typing.final`` decorator to a class indicates to mypy
+that it can't be subclassed. Mypy doesn't provide any additional features for
```

Comment:
Discuss why users might want to declare a class as final. Here are some ideas:

* A class wasn't designed to be subclassed. Perhaps subclassing does not work as expected, or it's error-prone.
* You want to retain the freedom to arbitrarily change the class implementation in the future, and these changes might break subclasses.
* You believe that subclassing would make code harder to understand or maintain. For example, you may want to prevent unnecessarily tight coupling between base classes and subclasses.